kratos/utilities/function_parser_utility.h

```diff
@@ -0,0 +1,266 @@
+    ///@{
+
+    /// The index type definition
+    typedef std::size_t IndexType;
+
+    /// Counted pointer of GenericFunctionUtility
+    KRATOS_CLASS_POINTER_DEFINITION(GenericFunctionUtility);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /**
+     * @brief Default constructor
+     * @param rFunctionBody The string defining the function
+     * @param LocalSystem The parameters defining the local system
+     */
+    GenericFunctionUtility(
+        const std::string& rFunctionBody,
+        Parameters LocalSystem = Parameters{}
+        );
+
+    ///Copy constructor
+    GenericFunctionUtility(GenericFunctionUtility const& rOther);
+
+    /// Destructor.
+    ~GenericFunctionUtility();
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /**
+     * @brief This method clones the current function instance
+     * @return A clone of the current class
+     */
+    Pointer Clone()
+    {
+        return Kratos::make_shared<GenericFunctionUtility>(*this);
+    }
+
+    /**
+     * @brief This method returns if it depends on space
+     * @return True if it uses the local system, false otherwise
+     */
+    bool UseLocalSystem()
+    {
+        return mUseLocalSystem;
+    }
+
+    /**
+     * @brief This method returns if it depends on space
+     * @return True if it depends on space, false otherwise
+     */
+    bool DependsOnSpace()
+    {
+        return mDependsOnSpace;
+    }
+
+    /**
+     * @brief This method returns the function body
+     * @return The function body
+     */
+    std::string FunctionBody();
+
+    /**
+     * @brief This method rotates and calls the evaluation function
+     * @param x The x coordinate
+     * @param y The y coordinate
+     * @param z The z coordinate
+     * @param t The time variable
+     * @param X The initial x coordinate
+     * @param Y The initial y coordinate
+     * @param Z The initial z coordinate
+     */
+    double RotateAndCallFunction(
+        const double x,
+        const double y,
+        const double z,
+        const double t,
+        const double X = 0.0,
+        const double Y = 0.0,
+        const double Z = 0.0
+        );
+
+    /**
+     * @brief This calls the evaluation function
+     * @param x The x coordinate
+     * @param y The y coordinate
+     * @param z The z coordinate
+     * @param t The time variable
+     * @param X The initial x coordinate
+     * @param Y The initial y coordinate
+     * @param Z The initial z coordinate
+     */
+    double CallFunction(
+        const double x,
+        const double y,
+        const double z,
+        const double t,
+        const double X = 0.0,
+        const double Y = 0.0,
+        const double Z = 0.0
+        );
+
+    ///@}
+
+private:
+
+    ///@name Member Variables
+    ///@{
+
+    std::unordered_map<std::string, double>  mNameSpace;   /// The variables considered on the function
+    mu::Parser mParser;                                    /// The function parser
```

Comment:
is it expensive to create new parsers?
If not then I would refactor such that the include of Muparser can go to the Cpp

otherwise we include it often, would be better not to do that

Reply:
Do you mean to use pointers and forward declaration?

Reply:
No
I mean not to make this a member at all and only create it when needed

Reply:
But it is the essence of the class, cannot work without it. In my local changes aI alredy did the forward delcaration do you mind?

Reply:
yes I know it is the essence
Thats why I asked if it is possible/expensive to create this on the fly to avoid the member
=> this way it can live entirely in the cpp and doesn't roam around the Kratos namespace

we usually avoid forward declarations

Reply:
Any suggestion to measure cost of creation?, i wthink I prefer the member variable